add github push task, using util function in pipeline/utils.

mvashishtha opened the pull request:
I also added a fake repository for testing in test/fake-repos.

Once we had extended the Python and Java GRPC client pipelines in artman to optionally publish generated code, we wanted to extend the Go pipeline to do the same. For Go the standard practice for publishing is simply to upload code to a GitHub repository. Since the task of uploading generated code to a GitHub repository might be useful for other languages as well, we want to add a task that will upload a folder to a GitHub repository (independent of the particular language). Code similar to this will also be used for publishing API protos from google3 to GitHub.  

This commit uses the GitHub API's JSON over HTTP methods to create and push the commit, because libraries that wrap the GitHub API did not provide the required methods.

The implementation follows these steps ([suggested on the API page](https://developer.github.com/v3/git/)), manipulating git internals to publish the output folder:
1. get the current commit object
2. retrieve the tree it points to
3. retrieve the content of the blob object that tree has for that particular file path
4. change the content somehow and post a new blob object with that new content, getting a blob SHA back
5. post a new tree object with that file path pointer replaced with your new blob SHA getting a tree SHA back
6. create a new commit object with the current commit SHA as the parent and the new tree SHA, getting a commit SHA back
7. update the reference of your branch to point to the new commit SHA

geigerj commented:
The new fake repo is huge... (is it all of artman?). Can we test against a tiny fake repo containing just one or a few files, each file of only one or a few lines?

jmuk commented:
Why don't we simply invoke git command instead of using github APIs?

ethanbao commented:
@jmuk Using git command is one option that I once discussed with Mehash before he started implementing. Using git definitely will work for artman but not as simple as maven/pypi is, because the former would require steps like git clone/add/commit/push instead of a single publish step. I would definitely agree that direct use of git command brings lots of benefits. On the other hand, the use of API/library gives us better control and error handling. In addition, the same code will be reused in some other gapic projects where the runtime won't be able to have access to git binary.

mvashishtha commented:
@geigerj Yeah, it's all of artman. I agree that we don't need something so big for testing. I'll make the repo smaller on my next commit.

mvashishtha commented:
@geigerj @ethanbao @jmuk  PTAL. Updated the commit message and made the fake repo much smaller.

jmuk commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+# TODO: Have this function also upload symlinks to the repository
+# according to GitHub protocol
+def _build_git_tree(curr_dir, username, password, owner, repo):
+    """Create a git tree from curr_dir for the given GitHub repo"""
```
I think this can be written in a simpler way by using os.walk (see https://docs.python.org/2/library/os.html#os.walk)

mvashishtha replied:
We discussed this offline-- we decided that because its subtrees have to be created before the tree object is created, os.walk is not appropriate here.

jmuk commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+    branch_item = _get_request_content(requests.get,
+                                       '{0}/repos/{1}/{2}/'
+                                       'branches/{3}'.format(_API_URL, owner,
+                                                             repo, branch))
```
I like to make a linebreak before 'requests.get', that's better in this case -- such as:
```
branch_item = _get_request_content(
    request.get,
    '{0}/repos/{1}/{2}/branches/{3}'.format(_API_URL, owners, repo, branch))
```

jmuk replied:
Optional: <_API_URL>/repos/<owner>/<repo> is common among this file and other utilities.
I recommend to assign it to a variable, like `base_url = '{0}/repos/{1}/{2}/'.format(_API_URL, owner, repo)` at the beginning of this function, and share it among other invocations.

mvashishtha replied:
Both-- done.

jmuk commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+"""Utility functions related to GitHub interaction. Uses the GitHub API with
+reference https://developer.github.com/v3/"""
+
+# How to update requirements so that these imports are possible?
```
Remove this line

mvashishtha replied:
Done.

ethanbao commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -56,3 +58,15 @@ def execute(self, repo_url, username, password, publish_env,
+
+    def execute(self, owner, branch, username, password, publish_env,
+                final_repo_dir, message):
+        github_utils.push_dir_github(final_repo_dir, username, password, owner,
```
Name the first param to "dir_to_push", and use "rebind" (http://docs.openstack.org/developer/taskflow/atoms.html) to map "final_repo_dir" to "dir_to_push" when creating the github publish task. By doing this, we don't force the final_repo_dir to be the input/output name.

mvashishtha replied:
Done.

ethanbao commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -56,3 +58,15 @@ def execute(self, repo_url, username, password, publish_env,
+
+
+class GitHubPushTask(task_base.TaskBase):
+    """Uploads to a GitHub repository"""
```
Let's add details about this task, and emphasize that this task adds and pushes the files in the given directory to a given github repo. It doesn't handle merge and won't delete files (if missing in the local repo) from the remote repo.

mvashishtha replied:
Done.

ethanbao commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+                  os.listdir(a_dir))
+
+
+def _get_request_content(req_func, url, req_data=None, username=None,
```
I feel it more readable if we replace "req_func" param with an enum param (http_method, which can be just GET and POST in this case). For the method name, can we call it "_exec_request", and mention in the comment that this method returns the json response if the request succeeds. Otherwise an exception will be raised.

mvashishtha replied:
Done.

ethanbao commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+                                  password, owner, repo)
+        tree_els.append({'path': subdir,
+                         'mode': _TREE_EL_MODES['tree'],
+                         'type': 'tree', 'sha': subtree['sha']})
```
nit: new line before 'sha'.

mvashishtha replied:
Done.

ethanbao commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+        tree_els.append({'path': subdir,
+                         'mode': _TREE_EL_MODES['tree'],
+                         'type': 'tree', 'sha': subtree['sha']})
+    for afile in _get_entries(_is_file, curr_dir):
```
nit: let's name "afile" to "file_entry" or "file".

mvashishtha replied:
"afile" no longer used

ethanbao commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+    return os.path.isfile(el) and os.access(el, os.X_OK)
+
+
+def _get_entries(entry_type_func, a_dir):
```
nit: s/a_dir/dir

mvashishtha replied:
dir is a built-in function in python, so I didn't want to override it.

ethanbao commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,137 @@
+                        'Reason: {1}'.format(url, attempt.content))
+
+
+def _build_git_tree_blob_item(file_or_exec, blob_mode, curr_dir, username,
```
Would that be any difference if the input is file vs executable?

mvashishtha replied:
GitHub requires blobs to have different modes if they are executables rather than files.

mvashishtha commented:
@ethanbao @jmuk PTAL. 

I added a commit with responses to your initial comments. I made the fake test directories even smaller. I also made the test code simpler. 

I also made a change in functionality after a discussion with Ethan. The previous commit created an entirely new tree, and deleted files that were in the remote repo but not in the local repo. This commit leaves such files alone. It will still overwrite files in the remote repo with content from the local files, however. The commit message has been updated to reflect this.

jmuk commented on test/utils/test_github_utils.py:
```diff
@@ -0,0 +1,145 @@
+    github_utils.push_dir_github(_FAKE_LOCAL, _USERNAME, _PASSWORD, _OWNER,
+                                 _REPO, _BRANCH, _MESSAGE)
+    mock_args = mock_exec_request.call_args_list
+    assert mock_args == EXPECTED_CALL_ARGS
```
nit: `mock_exec_request.assert_has_calls(EXPECTED_CALL_ARGS)` would be fine.

jmuk commented:
LGTM

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+
+def _get_immediate_subdirs(a_dir):
+    return filter(lambda x: os.path.isdir(os.path.join(a_dir, x)),
```
optional: I find list comprehension more readable here and below:

`[subdir for subdir in os.listdir(a_dir) if os.path.isdir(os.path.join(a_dir, subdir))]`

mvashishtha replied:
Done for both functions

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+def _exec_request(http_method, url, req_data=None, username=None,
+                  password=None):
+    """Make an HTTP request with given data and (username, password) tuple.
+    Return json response if request succeeds, otherwise raise an exception"""
```
Prefer proper google-style doc formatting for return value and exception raising: https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments

mvashishtha replied:
Should I list the args in the docstring as well?

mvashishtha replied:
Done

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+                  'executable': u'100755',
+                  'tree': u'040000',
+                  'symlink': u'120000'}
+_REQUEST_FUNCS = {'GET': requests.get,
```
Not sure that this map is really useful -- it requires using string values (e.g., line 104) below where it would be simpler just to pass `requests.XXX` as a parameter directly.

mvashishtha replied:
Originally I was passing function objects directly but Ethan thought that enums would be more readable, and I thought that this approach with a dictionary would be more like using enums. I will talk with Ethan about this-- even if we use the Python enum class I think that passing the function objects themselves will be more readable.

mvashishtha replied:
Changed back to passing functions

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+                  password=None):
+    """Make an HTTP request with given data and (username, password) tuple.
+    Return json response if request succeeds, otherwise raise an exception"""
+    request_func = _REQUEST_FUNCS[http_method]
```
See comment above. Would prefer `http_method` were just in `[requests.get, requests.post, requests.patch]` so that you don't need to translate string params

mvashishtha replied:
Resolved above

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+    if attempt.ok:
+        return json.loads(attempt.content)
+    else:
+        raise Exception(
```
Instead of using the Exception base class, raise an appropriate subclass. It's fine to create one if it doesn't exist.

mvashishtha replied:
Changed to HTTPError

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+_API_URL = 'https://api.github.com'
+# file modes for different git file types
+_TREE_EL_MODES = {'file': u'100644',
```
Instead of dict, why not just `FILE_MODE`, `EXECUTABLE_MODE`, ...?

mvashishtha replied:
Done

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+        subtree = _extend_git_tree(
+            os.path.join(curr_dir, subdir), username, password, base_url,
+            subtree_sha)
+        tree_els.append({'path': subdir, 'mode': _TREE_EL_MODES['tree'],
```
See comment above about `_TREE_EL_MODES`

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+# TODO: Have this function also upload symlinks to the repository
+# according to GitHub protocol
+def _extend_git_tree(curr_dir, username, password, base_url,
```
This function is non-trivial -- could you please add a short docstring describing it, even though it's private?

mvashishtha replied:
Done

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+    if orig_tree_sha:
+        data_dict['base_tree'] = orig_tree_sha
+    if not tree_els:
+        raise Exception("Cannot create empty folder {0} in Git".format(
```
Don't raise base Exception -- pick or create an appropriate subclass.

mvashishtha replied:
This error case was moved to push_dir_github and raises a ValueError

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,132 @@
+
+def push_dir_github(output_dir, username, password, owner, repo, branch,
+                    message='Automated commit from artman'):
+    """ Push all content in output_dir to the given GitHub repo branch"""
```
extraneous space before docstring (`""" P` --> `"""P`)

mvashishtha replied:
Done

mvashishtha commented:
@geigerj @ethanbao PTAL. Responded to comments. Also made other changes (mentioned in commit message)

geigerj commented on test/utils/test_github_utils.py:
```diff
@@ -0,0 +1,147 @@
+    return []
+
+
+@mock.patch('pipeline.utils.github_utils._get_immediate_subdirs',
```
Don't mock the pipeline -- we want to test this. Instead mock at the lowest level you can -- here, I think that would be the `requests` library.

mvashishtha replied:
The function that I'm testing has to use os functions to walk through the specified local directory, and I have to mock the function for listing directory items so that I know what order the tree will be built in. What do you mean by mocking the pipeline?

geigerj replied:
I mean that it is better to do

```
@mock.patch('os.path.isdir')
@mock.patch('os.path.listdir')
@mock.patch('requests....')
```

instead of mocking the same code that you're testing -- these utility methods are untested if you're mocking them.

jmuk replied:
I agree that this does not mock the pipeline -- but I am not sure if mocking your own internal function is the best approach.

Probably it is better to mock a filesystem itself (with pyfakefs https://github.com/jmcgeheeiv/pyfakefs) and mock the HTTP functions.

geigerj replied:
Sorry if I was unclear -- by "mock the pipeline", I just meant mocking functions that are defined in the artman repo, as opposed to its dependencies.

jmuk replied:
Yeah, I see (I wrote my reply before I see your 2nd comment).  pyfakefs looks like the utility to mock `os.path` functions consistently.

mvashishtha replied:
Test now uses pyfakefs and mocks at os level.

geigerj commented on test/utils/test_github_utils.py:
```diff
@@ -0,0 +1,147 @@
+                                _USERNAME, _PASSWORD)]
+
+
+def _requests_side_effect(http_method, url, req_data=None, username=None,
```
As discussed offline, I recommend not representing the remote tree, because it requires a substantial amount of complexity in this function -- ideally, the test should have little logic of its own.

Instead, can we mock `requests` and explicitly specify each response to each request, with no reference to the remote tree as a whole?

mvashishtha replied:
Now explicitly specifying response to each request.

geigerj commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -56,3 +58,19 @@ def execute(self, repo_url, username, password, publish_env,
+
+
+class GitHubPushTask(task_base.TaskBase):
+    """Uploads local files to GitHub repository in a new commit. Won't
```
Nit: prefer to start docstrings with a single summary line, and give details in the following paragraph.

mvashishtha replied:
Changed docstrings to this pattern.

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,163 @@
+        json response to request
+
+    Raises:
+        HTTPError: HTTP Request fails
```
Use past tense here and elsewhere ("The HTTP request failed.")

mvashishtha replied:
Done

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,163 @@
+    Returns:
+        New tree object as a dictionary, if curr_dir has files
+        Otherwise, returns None
+    Raises:
```
Use a newline between `Returns:` and `Raises:` blocks

mvashishtha replied:
Done

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,163 @@
+    Returns:
+        None
+    Raises:
+        ValueError: curr_dir is empty
```
There's no parameter to this function called `curr_dir`

mvashishtha replied:
Changed to output_dir

mvashishtha commented:
@geigerj @jmuk PTAL. Tests now mock at os and requests level as suggested. I use a fake directory structure with pyfakefs, so I deleted the fake local repo. I also responded to Jacob's comments on the docstrings.

geigerj commented on pipeline/tasks/publish_tasks.py:
```diff
@@ -56,3 +58,21 @@ def execute(self, repo_url, username, password, publish_env,
+    remote copy of a file differs from local copy, overwrites with local copy.
+    Does not change files or folders that are in remote repo but not in local
+    repo.
+
```
Nit: no blank line before the closing `"""` (here and elsewhere)

mvashishtha replied:
Done

geigerj commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,165 @@
+    root_tree = _extend_git_tree(output_dir, username, password,
+                                 base_url, orig_tree_sha)
+    if not root_tree:
+        raise ValueError("Cannot push empty folder {0}".format(output_dir))
```
nit: prefer single quotes for strings for consistency with the rest of the codebase

mvashishtha replied:
Changed to single quotes

geigerj commented:
Two nits, then LGTM

geigerj commented on tox.ini:
```diff
@@ -16,7 +16,7 @@ commands = py.test --timeout=30 --cov-report html --cov-report=term --cov {toxin
 [testenv:pep8]
 deps = flake8
-commands = flake8 --max-complexity=8 pipeline --exclude='test/output' test
+commands = flake8 --max-complexity=8 pipeline --exclude='test/output','test/fake-repos' test
```
Do you still need this since you are now mocking the file system?

mvashishtha replied:
Not needed anymore, with fake file system. Removed this part.

ethanbao commented on pipeline/utils/github_utils.py:
```diff
@@ -0,0 +1,165 @@
+                             json.dumps(data_dict), username, password)
+
+
+def push_dir_github(output_dir, username, password, owner, repo, branch,
```
nit: push_dir_to_github, and rename the first param to dir.

mvashishtha replied:
Since dir is reserved name in python, renamed output_dir to local_dir. Renamed push_dir_github to push_dir_to_github.

ethanbao commented:
LGTM
